spec/std/openssl/ssl/hostname_validation_spec.cr

```diff
@@ -0,0 +1,54 @@
+require "spec"
+require "openssl/ssl/hostname_validation"
+
+describe OpenSSL::SSL::HostnameValidation do
+  describe "matches_hostname?" do
+    it "matches ipv4, ipv6" do
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("192.168.1.1", "192.168.1.1").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("192.168.1.2", "192.168.1.1").should be_false
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("::1", "::1").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("::1", "::2").should be_false
+    end
+
+    it "won't wildcard match against ip" do
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("*.168.1.1", "192.168.1.1").should be_false
+    end
+
+    it "skips trailing dot" do
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("example.com.", "example.com").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("example.com", "example.com.").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?(".example.com", "example.com").should be_false
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("example.com", ".example.com").should be_false
```

Comment:
"trailing or leading" then? Also I remember something about `.example.com` being valid for the entire zone, at least any labels below it, not sure if it includes the given apex too. While probably very rare in practice though.